Doc/reference/datamodel.rst

```diff
@@ -2037,12 +2037,15 @@ Basic customization
    .. index:: single: __len__() (mapping object method)
 
    Called to implement truth value testing and the built-in operation
-   ``bool()``; should return ``False`` or ``True``.  When this method is not
+   :func:`bool`; should return ``False`` or ``True``.  When this method is not
    defined, :meth:`~object.__len__` is called, if it is defined, and the object is
    considered true if its result is nonzero.  If a class defines neither
    :meth:`!__len__` nor :meth:`!__bool__`, all its instances are considered
    true.
 
+   Two successive calls to :meth:`!__bool__` on the same object must
+   return same value.
```

Comment:
```suggestion
   Note: Python interpreters are permitted to assume that two successive calls
   to  :meth:`!__bool__` on the same object in the same thread will return the same
   value. While this expected invariant is not explicitly enforced, failing to abide
   by it will result in implementation dependent runtime behaviour. This
   implementation dependent behaviour may also be encountered when mutable
   objects are shared across threads without appropriate synchronization.
```

Reply:
I'm unsure if this needs to be explicitly documented as the behavior is dependent on the user's choice of implementation.

Reply:
> is dependent on the user's choice of implementation.

Rather on bugs in the user code, if someone will implement ``__bool__()``, doing crazy things (see issue).  Added docs say that implementation may assume certain behaviour from the user code, just as for the ``__hash__()`` method (same hash value for equal objects - the invariant, which we also can't enforce, user code might break this).

PS: I think that part of discussion rather belongs to the issue thread, which has some other arguments on why we want document this.  See e.g. [this](https://github.com/python/cpython/issues/124285#issuecomment-2378222286).

Reply:
Should the note warn against side effects in general (like I/O), not just mutation?

Reply:
> Should the note warn against side effects in general (like I/O), not just mutation?

Side effects, including in fact object mutation - are fine, unless they break idempotence.

But I think we don't loose something practically relevant if just forbid mutation of any objects in ``__bool__()`` (a shortened version of @ncoghlan suggestion):
```suggestion
   Calls to :meth:!__bool__` shouldn't mutate any objects.
```